Deliver secondary events first

heplesser opened the pull request:
@akorgor @JesusEV This PR moves delivery of secondary events before delivery of primary events.

heplesser committed:
Fix problem with readout signal propagation

heplesser committed:
Move gradient update to eprop neurons and remove neuron-specific synapses

heplesser committed:
Deliver secondary before primary events

akorgor commented on models/eprop_iaf_psc_delta.cpp:
```diff
@@ -63,6 +63,8 @@ RecordablesMap< eprop_iaf_psc_delta >::create()
   insert_( names::V_m, &eprop_iaf_psc_delta::get_V_m_ );
 }
 
+} // namespace nest
```
I like that change! Should we maybe get rid of the `namespace nest` altogether (of course by explicitly adding `nest::` wherever needed)?

heplesser replied:
We need the `namespace nest { ... }` for this part, if I remember earlier experimentation correctly.

akorgor replied:
It works, if `nest::` is inserted in the following way:
```cpp
void
nest::register_eprop_iaf_psc_delta( const std::string& name )
{
  register_node_model< nest::eprop_iaf_psc_delta >( name );
}

/* ----------------------------------------------------------------
 * Recordables map
 * ---------------------------------------------------------------- */

nest::RecordablesMap< nest::eprop_iaf_psc_delta > nest::eprop_iaf_psc_delta::recordablesMap_;

template <>
void
nest::RecordablesMap< nest::eprop_iaf_psc_delta >::create()
{
  insert_( nest::names::learning_signal, &eprop_iaf_psc_delta::get_learning_signal_ );
  insert_( nest::names::surrogate_gradient, &eprop_iaf_psc_delta::get_surrogate_gradient_ );
  insert_( nest::names::V_m, &eprop_iaf_psc_delta::get_V_m_ );
}
```

heplesser replied:
You are right, but we have this style in many other neuron models, I would leave fixing that to another day.

JesusEV commented on nestkernel/simulation_manager.cpp:
```diff
@@ -849,6 +849,15 @@ nest::SimulationManager::update_()
+          //
+          // Delivering secondary events first provides LearningSignalConnectionEvents
+          // to target neurons before spikes are transmitted through eprop synapses
+          // targeting the neurons, reducing the need for buffering of information.
```
I think the comment would be a little more precise in the following way:

"Delivering secondary events ahead of primary ones ensures that LearningSignalConnectionEvents reach the target neurons before any spikes are propagated via eprop synapses. This sequence safeguards the gradient computation mechanism from missing critical information from the timestep immediately preceding that of the arrival of the spike that triggers the weight update."

heplesser replied:
Fixed.

JesusEV commented on models/eprop_readout.cpp:
```diff
@@ -313,23 +313,22 @@ nest::eprop_readout::compute_error_signal_cross_entropy_loss( const long& lag )
-    double signal = e.get_weight() * e.get_coeffvalue( it ); // get_coeffvalue advances iterator
-    long t = delay + i;
-
-    if ( rport == READOUT_SIG - MIN_RATE_RECEPTOR )
-      B_.normalization_rates_.add_value( t, signal );
-    else if ( rport == TARGET_SIG - MIN_RATE_RECEPTOR )
-      B_.delayed_rates_.add_value( t, signal );
+    const double signal = e.get_weight() * e.get_coeffvalue( it ); // get_coeffvalue advances iterator
+    if ( rport == READOUT_SIG )
+    {
+      B_.normalization_rates_.add_value( i, signal );
+    }
+    else if ( rport == TARGET_SIG )
+    {
+      B_.delayed_rates_.add_value( i, signal );
+    }
```
To contrast Agnes' solution to the problem that regression is now broken by this change, I found:

```
  const long delay = e.get_delay_steps();

  long i = 0;
  auto it = e.begin();
  while ( it != e.end() )
  {
    const double signal = e.get_weight() * e.get_coeffvalue( it ); // get_coeffvalue advances iterator
    long t = delay + i;
    if ( rport == READOUT_SIG )
    {
      B_.normalization_rates_.add_value( i, signal );
    }
    else if ( rport == TARGET_SIG )
    {
      B_.delayed_rates_.add_value( t, signal );
    }

    ++i;
  }
```

heplesser replied:
@JesusEV So you write to entry `i` in `normalization_rates_` and to entry `i + delay` in `delayed_rates_`, don't you?

And all eprop work is based on the time resolution being 1 ms and all delays being 1 ms, so that all delays are equal to 1 time step, correct? I think this through.

BTW, could you send me a verification script also for regression?

heplesser committed:
Improved comment

Co-authored-by: JesusEV <espinozavalverd@uni-wuppertal.de>

heplesser reviewed (commented):
@JesusEV Thanks for your comments, see my replies below.

JesusEV commented:
@heplesser regarding "So you write to entry i in normalization_rates_ and to entry i + delay in delayed_rates_, don't you?" that was a temporary solution I implemented to make both regrssion and classification work before I became familiar with Agnes's approach, which appears to be more consistent with the methodd used in other handle functions. I added how I did it just for the record.

JesusEV commented:
@heplesser  regarding "And all eprop work is based on the time resolution being 1 ms and all delays being 1 ms, so that all delays are equal to 1 time step, correct? I think this through."

In our current implementation in NEST, yes, all delays are equal to the resolution, and equivalent to one time step.

However, the theoretical framework of eprop allows the delays between recurrent neurons to be different than 1 time-step, which is something that requires some adjustments to the learning rule (a generalization) and I consider that it is doable in NEST.

heplesser committed:
Merge branch 'eprop_feature' of github.com:jstapmanns/nest-simulator into deliver-secondary-first

heplesser committed:
Protect against running eprop with resolution or delay != 1

heplesser committed:
Enforce unit resolution and delay for eprop

heplesser commented:
@akorgor @JesusEV I have now fixed the readout buffer in here closely following the approach you had in the files you mailed me, i.e., without ring buffers. I added checks then that force resolution and delay to 1 ms. This required fixing some tests.

I also merged all of your changes into this PR and hope that tests will now go through here.

The regression script still claims to fail, but differences are just in the last decimals, so this looks like rounding errors which should not worry us:
```
 verification FAILED ! 

[101.96435699904158, 103.4667311262058, 103.34060707477168, 103.68024403768639, 104.41277574875247]
[101.96435699904154, 103.46673112620576, 103.34060707477164, 103.68024403768634, 104.41277574875241]
```

At the same time, the classification and regression test in the testsuite both passed even when one of your scripts truly failed. Could you check the tests again and make them tighter (as the scripts are)?

I will close #30 and #31 now, they are both included here.

heplesser committed:
Fix pylint problem

akorgor commented:
> @akorgor @JesusEV I have now fixed the readout buffer in here closely following the approach you had in the files you mailed me, i.e., without ring buffers. I added checks then that force resolution and delay to 1 ms. This required fixing some tests.
> 
> I also merged all of your changes into this PR and hope that tests will now go through here.
> 
> The regression script still claims to fail, but differences are just in the last decimals, so this looks like rounding errors which should not worry us:
> 
> ```
>  verification FAILED ! 
> 
> [101.96435699904158, 103.4667311262058, 103.34060707477168, 103.68024403768639, 104.41277574875247]
> [101.96435699904154, 103.46673112620576, 103.34060707477164, 103.68024403768634, 104.41277574875241]
> ```
> 
> At the same time, the classification and regression test in the testsuite both passed even when one of your scripts truly failed. Could you check the tests again and make them tighter (as the scripts are)?
> 
> I will close #30 and #31 now, they are both included here.

Why did you decide against your solution of the ring buffer? For us, that one was working with our slight modifications. So, the regression tutorial script verification was also successful as well as the tests.

JesusEV commented:
> @akorgor @JesusEV I have now fixed the readout buffer in here closely following the approach you had in the files you mailed me, i.e., without ring buffers. I added checks then that force resolution and delay to 1 ms. This required fixing some tests.
> 
> I also merged all of your changes into this PR and hope that tests will now go through here.
> 
> The regression script still claims to fail, but differences are just in the last decimals, so this looks like rounding errors which should not worry us:
> 
> ```
>  verification FAILED ! 
> 
> [101.96435699904158, 103.4667311262058, 103.34060707477168, 103.68024403768639, 104.41277574875247]
> [101.96435699904154, 103.46673112620576, 103.34060707477164, 103.68024403768634, 104.41277574875241]
> ```
> 
> At the same time, the classification and regression test in the testsuite both passed even when one of your scripts truly failed. Could you check the tests again and make them tighter (as the scripts are)?
> 
> I will close #30 and #31 now, they are both included here.

@heplesser The eprop tests from the test suite continue to pass likely due to their use of downscaled networks. It is conceivable that the minuscule numerical differences (15th or 16th decimal places) observed in the tutorial script are too insignificant to manifest in the downsacled case. Consequently, I don't see a feasible way to further tighten those tests.

By the way, once everything stabilizes, we plan to remove the verifications from the tutorial scripts. The test suite will then serve as the sole source of validation checks.

akorgor commented:
I just tried the tutorials and the tests and everything passes. Maybe the results depend on the machine?

JesusEV commented:
> I just tried the tutorials and the tests and everything passes. Maybe the results depend on the machine?

I guess this is an instance of the kind of problems that can arise from comparing floating point numbers using the == operator. The differences are so small that probably already dwell in machine epsilon region. Probably a tolerance (a very small one) could be more appropriate here?

akorgor commented:
> > I just tried the tutorials and the tests and everything passes. Maybe the results depend on the machine?
> 
> I guess this is an instance of the kind of problems that can arise from comparing floating point numbers using the == operator. The differences are so small that probably already dwell in machine epsilon region. Probably a tolerance (a very small one) could be more appropriate here?

Yes, this is a good idea.

akorgor commented on testsuite/pytests/.#foo:
```diff
@@ -0,0 +1 @@
+plesser@Hanss-MacBook-Pro.local.32815
```
Is this supposed to be here?

heplesser replied:
Thanks for spotting this and the other file, fixed now!

akorgor commented on testsuite/pytests/foo:
```diff
@@ -0,0 +1,5 @@
+    | install.share.nest.testsuite.pytests.sli2py_recording.test_multimeter_stepping.test_multimeter_stepping[eprop_iaf_psc_delta]
```
Is this supposed to be here?

akorgor commented on testsuite/pytests/test_labeled_synapses.py:
```diff
@@ -56,7 +58,7 @@ def default_network(self, syn_model):
         self.urbanczik_synapses = ["urbanczik_synapse", "urbanczik_synapse_lbl", "urbanczik_synapse_hpc"]
 
-        self.eprop_synapses = ["eprop_synapse", "eprop_synapse_lbl", "eprop_synapse_hpc"]
+        self.eprop_synapses = ["eprop_synapse", "eprop_synapse_hpc", "eprop_synapse_lbl"]
```
```suggestion
        self.eprop_synapses = ["eprop_synapse", "eprop_synapse_lbl", "eprop_synapse_hpc"]
```

akorgor commented on pynest/examples/eprop_plasticity/eprop_supervised_classification.py:
```diff
@@ -594,7 +592,7 @@ def get_weights(pop_pre, pop_post):
 # pylint: disable-next=E1121,C0301
 readout_signal = readout_signal.reshape(n_out, n_iter, n_batch, steps["sequence"])[:, :, :, -steps["learning_window"] :]
-# pylint: disable-next=E1121,C0301
+# pylint: disable-next-=E1121,C0301
```
```suggestion
# pylint: disable-next=E1121,C0301
```

heplesser commented:
> > ```
> >  verification FAILED ! 
> > 
> > [101.96435699904158, 103.4667311262058, 103.34060707477168, 103.68024403768639, 104.41277574875247]
> > [101.96435699904154, 103.46673112620576, 103.34060707477164, 103.68024403768634, 104.41277574875241]
> > ```
> @heplesser The eprop tests from the test suite continue to pass likely due to their use of downscaled networks. It is conceivable that the minuscule numerical differences (15th or 16th decimal places) observed in the tutorial script are too insignificant to manifest in the downsacled case. Consequently, I don't see a feasible way to further tighten those tests.
> 
> By the way, once everything stabilizes, we plan to remove the verifications from the tutorial scripts. The test suite will then serve as the sole source of validation checks.

These tiny differences are can most likely be explained by numerical issues and are nothing to worry about. When we had differences before, they typically popped up in the 3rd or 4th decimal at the latest. So I think it is most sensible to use the Numpy or Pytest functions for approximate comparisons with a relative tolerance around 10^-7.

Concerning the tightening: how about using the network configuration from the test scripts that showed the error in `test_eprop_plasticity.py`? The scripts executed quite fast and were clearly big enough for an error to show up, while the configurations used in in test_eprop... are too small.

heplesser committed:
Remove accidentally commited files

heplesser committed:
Update testsuite/pytests/test_labeled_synapses.py

Co-authored-by: Agnes Korcsak-Gorzo <40828647+akorgor@users.noreply.github.com>

heplesser committed:
Update pynest/examples/eprop_plasticity/eprop_supervised_classification.py

Co-authored-by: Agnes Korcsak-Gorzo <40828647+akorgor@users.noreply.github.com>

akorgor commented:
> > > ```
> > >  verification FAILED ! 
> > > 
> > > [101.96435699904158, 103.4667311262058, 103.34060707477168, 103.68024403768639, 104.41277574875247]
> > > [101.96435699904154, 103.46673112620576, 103.34060707477164, 103.68024403768634, 104.41277574875241]
> > > ```
> > 
> > 
> > @heplesser The eprop tests from the test suite continue to pass likely due to their use of downscaled networks. It is conceivable that the minuscule numerical differences (15th or 16th decimal places) observed in the tutorial script are too insignificant to manifest in the downsacled case. Consequently, I don't see a feasible way to further tighten those tests.
> > By the way, once everything stabilizes, we plan to remove the verifications from the tutorial scripts. The test suite will then serve as the sole source of validation checks.
> 
> These tiny differences are can most likely be explained by numerical issues and are nothing to worry about. When we had differences before, they typically popped up in the 3rd or 4th decimal at the latest. So I think it is most sensible to use the Numpy or Pytest functions for approximate comparisons with a relative tolerance around 10^-7.
> 
> Concerning the tightening: how about using the network configuration from the test scripts that showed the error in `test_eprop_plasticity.py`? The scripts executed quite fast and were clearly big enough for an error to show up, while the configurations used in in test_eprop... are too small.

Yes, this is possible. The tests are simply the tutorial tasks downscaled to 10 % (in network size and task length), controlled by a dedicated parameter `scale`. We scaled it down, just for saving execution time, but the full-scale tasks should be still fast.
